### PR TITLE
First pass at course discovery search - SOL-280

### PR DESCRIPTION
**Pull request by martynjames**
For first review (@mattdrayer) This is the search-side changes to enable course discovery

### CODE REVIEW

**Comment by martynjames** [comment]
@dsego - This is a first pass at the search code that you'll need in place to expose the course_discovery endpoint.

File: search/api.py [hunk 1]
```diff
@@ -30,18 +36,41 @@ def perform_search(
...
     return results
+
+
+def course_discovery_search(search_term=None, size=20, from_=0):  # , **kwargs):
+    """
+    Course Discovery activities against the search engine index of course details
+    """
+    searcher = SearchEngine.get_search_engine(getattr(settings, "COURSEWARE_INDEX_NAME", "courseware_index"))
+    if not searcher:
+        raise NoSearchEngine("No search engine specified in settings.SEARCH_ENGINE")
```
Comment (mattdrayer) [inline search/api.py]: Nit: I usually see exceptions suffixed with "Exception" or "Error"

File: search/views.py [hunk 2]
```diff
@@ -11,13 +11,114 @@
...
+        "search_string" (optional) - text with which to search for courses
+        "page_size" (optional)- how many results to return per page (defaults to 20, with maximum cutoff at 100)
+        "page_index" (optional) - for which page (zero-indexed) to include results (defaults to 0)
+    """
+    results = {
+        "error": _("Nothing to search")
+    }
+    status_code = 500
+
+    try:
```
Comment (mattdrayer) [inline search/views.py]: In general I think it's better to wrap your try/except protection as tightly as possible around the piece of code you're concerned with, versus doing a lot of different things within the try block.

**Comment by mattdrayer** [comment]
Made an initial pass -- just a couple small things and maybe reworking one try/except block a bit

**Comment by martynjames** [comment]
Thanks @mattdrayer - I tried making lots of tighter try/except blocks, but when it comes down to it, this is an external entry point to the system, so I kept a single one around the operation. However, I did try to clean it up a little by moving some of the operations to their own functions - hopefully this will help.

File: search/urls.py [hunk 3]
```diff
@@ -12,4 +12,5 @@
     '',
     url(r'^$', views.do_search, name='do_search'),
     url(r'^{}$'.format(COURSE_ID_PATTERN), views.do_search, name='do_search'),
+    url(r'^course_discovery$', views.course_discovery, name='course_discovery'),
```
Comment (dsego) [inline search/urls.py]: Maybe end with a slash just for consistency with r'^search/'.

**Pull request merged**